proposed/http-client/http-client.md

```diff
@@ -21,9 +21,9 @@ implementations. This would make libraries more stable since the reduced number
 dependencies and the likelihood to get in version conflicts is reduced.
 
 The second goal is that all HTTP clients should follow the [Liskov substitutions principle][Liskov].
-This means that all clients should act the same when sending a request. By default a HTTP client
-should not follow redirect nor throw exceptions on HTTP responses with status 4xx or 5xx.
-
+This means that all clients should act the same when sending a request. A consuming
+library SHOULD expect a HTTP client to NOT follow redirects NOR throw exceptions based
+on HTTP response statuses.
```

Comment:
IMO, the difference in behavior (follow vs not follow) is problematic for exactly the same reason that throwing vs not throwing for certain HTTP status codes was problematic: it means that the client can *not* be safely substituted for another client implementation.

In other words, consuming code will *have* to assume that clients do *not* follow redirects, because they might not - in exactly the same way that clients would *have* to assume that all implementations throw.

I think this behavior needs to be consistent.

I'm not sure which I would prefer though, following or not following redirects. There are use-cases for both. Likely the most common use-case is following - but stipulating that clients *must* follow redirects means that the client abstraction can't be used at all for those use-cases where it *shouldn't* follow redirects.

So it becomes a question of supporting all use-cases (not following) vs convenience in the majority of use-cases (following).

That's a tough call. Likely this is why clients at large make this configurable.